File: exercism/csharp/space-age/SpaceAge.cs
using System.Collections.Generic;

namespace Exercism
{
    public class SpaceAge
    {
        public long Seconds { get; private set; }

        enum Planet {
            Earth,
            Mercury,
            Venus,
            Mars,
            Jupiter,
            Saturn,
            Uranus,
            Neptune
        };

        int earthSeconds = 31557600;

        Dictionary<Planet, double> secondsPerYear;

        public SpaceAge(long seconds)
        {
            Seconds = seconds;
            secondsPerYear = new Dictionary<Planet, double> {
                { Planet.Earth, 1.0 * earthSeconds },
                { Planet.Mercury, 0.2408467 * earthSeconds },
                { Planet.Venus, 0.61519726 * earthSeconds },
                { Planet.Mars, 1.8808158 * earthSeconds },
                { Planet.Jupiter, 11.862615 * earthSeconds },
                { Planet.Saturn, 29.447498 * earthSeconds },
                { Planet.Uranus, 84.016846 * earthSeconds },
                { Planet.Neptune, 164.79132 * earthSeconds }
            };
        }

        public double OnEarth() { return ageOnPlanet(Planet.Earth); }
        public double OnMercury() { return ageOnPlanet(Planet.Mercury); }
        public double OnVenus() { return ageOnPlanet(Planet.Venus); }
        public double OnMars() { return ageOnPlanet(Planet.Mars); }
        public double OnJupiter() { return ageOnPlanet(Planet.Jupiter); }
        public double OnSaturn() { return ageOnPlanet(Planet.Saturn); }
        public double OnUranus() { return ageOnPlanet(Planet.Uranus); }
        public double OnNeptune() { return ageOnPlanet(Planet.Neptune); }

        private double ageOnPlanet(Planet planet)
        {
            return Seconds / secondsPerYear[planet];
        }
    }
}

File: exercism/cpp/space-age/space_age.cpp
#include "space_age.h"

const long EARTH_SECONDS = 31557600;

float space_age::space_age::on_earth() const {
  return _seconds / (1.0 * EARTH_SECONDS);
}

float space_age::space_age::on_mercury() const {
  return _seconds / (0.2408467 * EARTH_SECONDS);
}

float space_age::space_age::on_venus() const {
  return _seconds / (0.61519726 * EARTH_SECONDS);
}

float space_age::space_age::on_mars() const {
  return _seconds / (1.8808158 * EARTH_SECONDS);
}

float space_age::space_age::on_jupiter() const {
  return _seconds / (11.862615 * EARTH_SECONDS);
}

float space_age::space_age::on_saturn() const {
  return _seconds / (29.447498 * EARTH_SECONDS);
}

float space_age::space_age::on_uranus() const {
  return _seconds / (84.016846 * EARTH_SECONDS);
}

float space_age::space_age::on_neptune() const {
  return _seconds / (164.79132 * EARTH_SECONDS);
}
